### ISSUE TITLE
in-browser coverage numbers

### ISSUE THREAD

**Issue by boneskull** [mochajs/mocha#2883]
We have coverage numbers, but only in Node.js.  I'd like to see the coverage numbers against tests run with Karma.  

Gathering coverage from Karma seems fraught with peril.

I would entertain other suggestions to improve how we run tests in a browser environment.  My only requirement is that the tests run in a strict browser-like environment--`jsdom` shenanigans or Electron won't do.

**Comment by ScottFreeCode** [comment]
I've been working on this in the `multi-coverage` branch.

Getting Coveralls to take reports from multiple build jobs in the same build turned out to be way easier than I thought. I wasted so many hours looking for a more complex solution because I didn't initially realize I could just apply `COVERAGE=true` to the whole build. That alone should increase the coverage somewhat as it will show branches taken in different Node versions; would be nice to get more such branches covered though (Windows, whatever other terminals we're handling).

After that, I looked into adding coverage to the browser runs. Here's what I found out.

- `karma-coverage` can instrument the code and report on coverage for instrumented code, *however*...
- `karma-browserify` doesn't look at the results of other preprocessors, *therefore*...
- `browserify-istanbul` should be used to make `karma-browserify` instrument the code, which can then be reported upon by `karma-coverage`.
- `browserify-istanbul` mysteriously doesn't run on anything going into the bundle that isn't in Karma's `files` list and given `['browserify']` on Karma's `preprocessors` list, so we have to add the relevant patterns to those lists even if they shouldn't be served directly.
- After I got all that figured it, it finally worked! ...Until I tried to clean up the other false starts I'd had, use NYC as described below, and consolidate the code behind `if (env.COVERAGE)`; then something changed and caused the current "bundle error" problems -- but I haven't been able to figure out what.
- It is possible to override the coverage implementation for both `browserify-istanbul` and `karma-coverage`; this can be used to make `browserify-istanbul` use the same version of Istanbul as `karma-coverage`, or to make both use the same version of Istanbul as our Node CLI NYC. (This *seems* to work ok except for the above-mentioned issue, and removing it does not seem to resolve the above-mentioned issue.)

Feel free to pull it down locally and experiment to confirm or deny any of the above and try to fix the bundle errors.

**Comment by ScottFreeCode** [comment]
(I should add that the `multi-coverage` branch is based off the CI fixes branch, for obvious reasons; the diff from master should get to be simpler once the CI fixes branch is merged, presumably...)

**Comment by ScottFreeCode** [comment]
Found and what I'd screwed up; that branch is working now (modulo Sauce flakes), so I've submitted PR #2886 for it.
